### PR TITLE
Text to speech block in app lab and game lab

### PR DESCRIPTION
**Pull request by JillianK**
Added a text to speech block to app lab and game lab that takes in text, output voice gender, and language (currently only english is supported). The block is under UI in app lab and World in game lab. It defaults to female and en-us if given invalid input. 
In app lab: 
Screen Shot 2020-05-20 at 4 39 39 PM

In game lab:
Screen Shot 2020-05-20 at 4 42 34 PM

as a block:
Screen Shot 2020-05-20 at 4 42 02 PM

### Future work 
At some point, we need to go back and add support for more languages. 

## Links

- [spec](https://docs.google.com/document/d/1EUVGzgSFprrwp9A_fEceidL3bhkTP12kxvEA_SfFSm0/edit)

## Testing story
There's a test for the arguments of play speech which I think should be sufficient (since we aren't pausing or stopping speech) but let me know if there's something else we need.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: .drone.yml
Comment (ajpal) [inline .drone.yml]: nit: intentional?

File: apps/src/Sound.js
Comment (ajpal) [inline apps/src/Sound.js:346]: why `false` not `undefined` or `null`?

File: apps/src/Sound.js
Comment (ajpal) [inline apps/src/Sound.js]: What's the purpose of having this line, and the one below after you check `if (file)` ?

File: apps/src/Sound.js
Comment (JillianK) [reply apps/src/Sound.js:346]: I was following the style of getPlayableFile (the function above). Should I switch to undefined?

File: apps/src/Sound.js
Comment (JillianK) [reply apps/src/Sound.js]: oops, I meant to remove that

File: apps/src/lib/util/audioApi.js
Comment (ajpal) [inline apps/src/lib/util/audioApi.js]: nit: `!!false` is just `false`, right?

File: apps/src/lib/util/audioApi.js
Comment (ajpal) [inline apps/src/lib/util/audioApi.js]: nit: remove?

File: apps/src/lib/util/audioApi.js
Comment (JillianK) [reply apps/src/lib/util/audioApi.js]: I guess? I was following the style of playSound (2 functions up), do you know why they did !!?

File: apps/src/lib/util/audioApi.js
Comment (JillianK) [reply apps/src/lib/util/audioApi.js]: any recommendations on what I should do with the error? I feel a little weird about letting the text to speech fail silently

File: apps/src/lib/util/audioApiDropletConfig.js
Comment (Hamms) [inline apps/src/lib/util/audioApiDropletConfig.js]: I'd suggest we display to the user the common name of the language, rather than the language code. Like what we do for the language dropdown in the footer

File: apps/src/lib/util/audioApi.js
Comment (Hamms) [reply apps/src/lib/util/audioApi.js]: `!!` is used to cast something to a boolean that might not be; I suspect it's used above to handle cases where `opts.loop` is undefined

Comment (JillianK) [reply apps/src/lib/util/audioApi.js]: ohh I see, thanks for explaining! I'll make the switch to just false

**Comment by jmkulwik** [comment]
IMO: Since we only support English, we shouldn't offer the option to change languages. That feels like a broken promise to me. We should add that option when we add more languages.

File: apps/src/Sound.js
Comment (jmkulwik) [inline apps/src/Sound.js]: Blanket catch statements are often hard to debug. Consider doing a console.warn here or some other sort of logging. (Also, I don't have an understanding of this use case, so if this makes more sense go ahead and keep it the way it is)

File: apps/src/Sound.js
Comment (jmkulwik) [inline apps/src/Sound.js]: wav does not work on IE11 - are we ok with this? (may be a product/curriculum question)

File: apps/test/unit/lib/util/audioApiTest.js
Comment (jmkulwik) [inline apps/test/unit/lib/util/audioApiTest.js]: nit: I generally prefer using strings that make sense in context as it gives the reader & future maintainer more knowledge about use patterns. Not required, but may be nice. i.e. for example set language to `en-us`

File: apps/src/Sound.js
Comment (JillianK) [reply apps/src/Sound.js]: Fixed by changing the format to mp3

File: apps/src/lib/util/audioApi.js
Comment (JillianK) [reply apps/src/lib/util/audioApi.js]: update: changed it to a console.warn

File: apps/i18n/common/en_us.json
Comment (jmkulwik) [inline apps/i18n/common/en_us.json]: nit: if removing language, we should also remove `in the specified language` here

**Comment by jmkulwik** [comment]
When I run this locally (Windows 10, Chrome), using the default block `playSpeech("Hello World!", "female");`, I get this error in the Applab debug console: 
image


EDIT: Wait a minute!! I wasn't auto-rebuilding after changes. Let me check that again.

**Comment by JillianK** [comment]
> When I run this locally (Windows 10, Chrome), using the default block `playSpeech("Hello World!", "female");`, I get this error in the Applab debug console:
> image
> 
> EDIT: Wait a minute. I wasn't auto-rebuilding after changes. Let me check that again.

If you want to run locally, you need keys from me

**Review by JillianK — COMMENTED** [review]
Just moved the block behind an experiment, text-to-speech-block, and removed it from the toolbox. To access it now, you need to enable the experiment and type in the block

File: apps/src/lib/util/audioApi.js
Comment (maddiedierker) [inline apps/src/lib/util/audioApi.js]: nit: this can be simplified to:

```
const forceHTML5 = window.location.protocol === 'file:';
```

File: apps/src/util/experiments.js
Comment (maddiedierker) [inline apps/src/util/experiments.js]: is this line purposeful? it looks unrelated

File: dashboard/app/helpers/levels_helper.rb
Comment (maddiedierker) [inline dashboard/app/helpers/levels_helper.rb]: it isn't very likely, but we should use our service region key when we're constructing this URI. then, if it changes, our URI doesn't break. something like

```
uri = URI.parse("https://#{CDO.azure_speech_service_region}.api.cognitive.microsoft.com/sts/v1.0/issueToken")
```

should work, but double-check that i didn't make any mistakes there

File: apps/src/util/experiments.js
Comment (JillianK) [reply apps/src/util/experiments.js]: it was a merge conflict I had to resolve

Comment (JillianK) [reply apps/src/util/experiments.js:31]: this was a merge conflict I had to resolve

Comment (jmkulwik) [reply apps/src/util/experiments.js]: You're probably good to go ahead and remove it now. Looks like it's been removed in staging.
https://github.com/code-dot-org/code-dot-org/commit/42005fd5906afbbb8d30533e20b2209bb2bcd4e0#diff-c0a24a87e19b1d7521e3b942f0fb1c21
https://github.com/code-dot-org/code-dot-org/blob/1d30a8cd4dcd4a50f4f3b0b44764e67e1cebbc83/apps/src/util/experiments.js#L30-L31

**Review by maddiedierker — APPROVED** [review]
🥇  really great work, jillian!!

**Comment by maddiedierker** [comment]
almost forgot -- before you merge this, let's add the API keys to each environment together. we can do this after today's standup

**Pull request merged**
